artiq/coredevice/urukul.py

```diff
@@ -124,6 +135,516 @@ def set_mu(self, ftw: TInt32):
+    and interact with the ProtoRev9 version of the CPLD.
+    """
+
+    # ProtoRev9 CFG configuration register bit offsets
+    CFG_OSK = 20
+    CFG_DRCTL = 24
+    CFG_DRHOLD = 28
+    CFG_IO_UPDATE = 32
+    CFG_MASK_NU = 36
+    CFG_CLK_SEL0 = 40
+    CFG_CLK_SEL1 = 44
+    CFG_SYNC_SEL = 41
+    CFG_RST = 42
+    CFG_IO_RST = 43
+    CFG_CLK_DIV = 45
+    CFG_ATT_EN = 47
+
+    @staticmethod
+    @portable
+    def urukul_cfg(
+        rf_sw,
+        led,
+        profile,
+        osk,
+        drctl,
+        drhold,
+        io_update,
+        mask_nu,
+        clk_sel,
+        sync_sel,
+        rst,
+        io_rst,
+        clk_div,
+        att_en,
+    ):
+        """Build Urukul CPLD configuration register"""
+        return (
+            (rf_sw << CFG_RF_SW)
+            | (led << CFG_LED)
+            | (profile << CFG_PROFILE)
+            | (osk << ProtoRev9.CFG_OSK)
+            | (drctl << ProtoRev9.CFG_DRCTL)
+            | (drhold << ProtoRev9.CFG_DRHOLD)
+            | (io_update << ProtoRev9.CFG_IO_UPDATE)
+            | (mask_nu << ProtoRev9.CFG_MASK_NU)
+            | ((clk_sel & 0x01) << ProtoRev9.CFG_CLK_SEL0)
+            | ((clk_sel & 0x02) << (ProtoRev9.CFG_CLK_SEL1 - 1))
+            | (sync_sel << ProtoRev9.CFG_SYNC_SEL)
+            | (rst << ProtoRev9.CFG_RST)
+            | (io_rst << ProtoRev9.CFG_IO_RST)
+            | (clk_div << ProtoRev9.CFG_CLK_DIV)
+            | (att_en << ProtoRev9.CFG_ATT_EN)
+        )
+
+    @kernel
+    def cfg_write(self, cpld, cfg: TInt64):
+        """Write to the configuration register.
+
+        See :func:`urukul_cfg` for possible flags.
+
+        :param cfg: 52-bit data to be written. Will be stored at
+            :attr:`cfg_reg`.
+        """
+        cpld.bus.set_config_mu(SPI_CONFIG, 24, SPIT_CFG_WR, CS_CFG)
+        cpld.bus.write(((cfg >> 28) & 0xFFFFFF) << 8)
+        cpld.bus.set_config_mu(SPI_CONFIG | spi.SPI_END, 28, SPIT_CFG_WR, CS_CFG)
+        cpld.bus.write((cfg & 0xFFFFFFF) << 4)
+        cpld.cfg_reg = cfg
+
+    @kernel
+    def sta_read(self, cpld) -> TInt64:
+        """Read the status register.
+
+        Use any of the following functions to extract values:
+
+            * :func:`urukul_sta_rf_sw`
+            * :func:`urukul_sta_smp_err`
+            * :func:`urukul_sta_pll_lock`
+            * :func:`urukul_sta_ifc_mode`
+            * :func:`urukul_sta_proto_rev`
+            * :func:`urukul_sta_drover`
+
+        :return: The status register value.
+        """
+        cpld.bus.set_config_mu(SPI_CONFIG | spi.SPI_INPUT, 24, SPIT_CFG_RD, CS_CFG)
+        cpld.bus.write(((cpld.cfg_reg >> 24) & 0xFFFFFF) << 8)
+        cpld.bus.set_config_mu(
+            SPI_CONFIG | spi.SPI_END | spi.SPI_INPUT, 28, SPIT_CFG_RD, CS_CFG
+        )
+        cpld.bus.write((cpld.cfg_reg & 0xFFFFFFF) << 4)
+        hi = cpld.bus.read()
+        lo = cpld.bus.read()
+        return (int64(hi) << 28) | lo
+
+    @kernel
+    def init(self, cpld):
```

Comment:
Inits for both 8 and 9 are missing proto_rev checks from STA. It's especially important to verify especially during the transitional period, and the simple side effect that a proto_rev check would also verify the presence of the device (e.g. would give a mismatch if EEM cables were mixed up)

Reply:
This was the route I tried originally, but not knowing which version doesn't work well with the compiler. That is, if I probe in `init()` for the `proto_rev`, once I have this information I then need to set which `self.cfg_reg` will be but the compiler wants to know beforehand what type this will be.  I don't see a way around this using the legacy compiler.  Additionally, `UrukulMonitor` relies on this information (I had forgotten to add this portion to the PR) so this will also need to be supplied beforehand as well.

Reply:
I understand that's a pain and there seems to be no way around it (could ``self.cfg_reg`` be a part of ``version``)? And even if you can't auto-detect and apply the correct driver, it should at very least verify that the user-provided version matches up with what Urukul is reporting.

Reply:
In regards to `self.cfg_reg` being part of the `self.version` instance, I would still need to pass in all the needed parameters from `__init__` to create the configuration register, so I am not really sure how this would help as I would need to create placeholders in `__init__` so that I could then use them within `init` depending on what probing would find, so currently I don't see a way around this.  Hopefully I understood what you were implying.

In regards to verifying that the user-provided versions match, yes I agree that would be a good addition and will add that.